diff --git a/app/vlstorage/main.go b/app/vlstorage/main.go
index 102c2b5..c560ea0 100644
--- a/app/vlstorage/main.go
+++ b/app/vlstorage/main.go
@@ -554,6 +554,10 @@ func (*Storage) MustAddRows(lr *logstorage.LogRows) {
 
 // RunQuery runs the given qctx and calls writeBlock for the returned data blocks
 func RunQuery(qctx *logstorage.QueryContext, writeBlock logstorage.WriteDataBlockFunc) error {
+	if externalStorage != nil {
+		return externalStorage.RunQuery(qctx, writeBlock)
+	}
+
 	if localStorage != nil {
 		// Optimize the query, which returns last N rows with the biggest timestamps,
 		// only at the leaf vlstorage nodes. There is no need in optimizing the query
@@ -576,6 +580,9 @@ func RunQuery(qctx *logstorage.QueryContext, writeBlock logstorage.WriteDataBloc
 //
 // If the filter isn't empty, then only the field names containing the filter substring are returned.
 func GetFieldNames(qctx *logstorage.QueryContext, filter string) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetFieldNames(qctx, filter)
+	}
 	if localStorage != nil {
 		return localStorage.GetFieldNames(qctx, filter)
 	}
@@ -588,6 +595,9 @@ func GetFieldNames(qctx *logstorage.QueryContext, filter string) ([]logstorage.V
 //
 // If limit > 0, then up to limit unique values are returned.
 func GetFieldValues(qctx *logstorage.QueryContext, fieldName, filter string, limit uint64) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetFieldValues(qctx, fieldName, filter, limit)
+	}
 	if localStorage != nil {
 		return localStorage.GetFieldValues(qctx, fieldName, filter, limit)
 	}
@@ -598,6 +608,9 @@ func GetFieldValues(qctx *logstorage.QueryContext, fieldName, filter string, lim
 //
 // If the filter is non-empty, then only the field names containing the filter substring are returned.
 func GetStreamFieldNames(qctx *logstorage.QueryContext, filter string) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetStreamFieldNames(qctx, filter)
+	}
 	if localStorage != nil {
 		return localStorage.GetStreamFieldNames(qctx, filter)
 	}
@@ -610,6 +623,9 @@ func GetStreamFieldNames(qctx *logstorage.QueryContext, filter string) ([]logsto
 //
 // If limit > 0, then up to limit unique stream field values are returned.
 func GetStreamFieldValues(qctx *logstorage.QueryContext, fieldName, filter string, limit uint64) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetStreamFieldValues(qctx, fieldName, filter, limit)
+	}
 	if localStorage != nil {
 		return localStorage.GetStreamFieldValues(qctx, fieldName, filter, limit)
 	}
@@ -620,6 +636,9 @@ func GetStreamFieldValues(qctx *logstorage.QueryContext, fieldName, filter strin
 //
 // If limit > 0, then up to limit unique streams are returned.
 func GetStreams(qctx *logstorage.QueryContext, limit uint64) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetStreams(qctx, limit)
+	}
 	if localStorage != nil {
 		return localStorage.GetStreams(qctx, limit)
 	}
@@ -630,6 +649,9 @@ func GetStreams(qctx *logstorage.QueryContext, limit uint64) ([]logstorage.Value
 //
 // If limit > 0, then up to limit unique streamIDs are returned.
 func GetStreamIDs(qctx *logstorage.QueryContext, limit uint64) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetStreamIDs(qctx, limit)
+	}
 	if localStorage != nil {
 		return localStorage.GetStreamIDs(qctx, limit)
 	}
@@ -642,6 +664,9 @@ func GetStreamIDs(qctx *logstorage.QueryContext, limit uint64) ([]logstorage.Val
 func DeleteRunTask(ctx context.Context, taskID string, timestamp int64, tenantIDs []logstorage.TenantID, f *logstorage.Filter) error {
 	logger.Infof("starting deleting logs for task_id=%q, filter=%q, tenantIDs=%s", taskID, f, tenantIDs)
 
+	if externalStorage != nil {
+		return externalStorage.DeleteRunTask(ctx, taskID, timestamp, tenantIDs, f)
+	}
 	if localStorage != nil {
 		return localStorage.DeleteRunTask(ctx, taskID, timestamp, tenantIDs, f)
 	}
@@ -653,7 +678,9 @@ func DeleteStopTask(ctx context.Context, taskID string) error {
 	logger.Infof("stopping delete task with task_id=%q", taskID)
 
 	var err error
-	if localStorage != nil {
+	if externalStorage != nil {
+		err = externalStorage.DeleteStopTask(ctx, taskID)
+	} else if localStorage != nil {
 		err = localStorage.DeleteStopTask(ctx, taskID)
 	} else {
 		err = netstorageSelect.DeleteStopTask(ctx, taskID)
@@ -666,6 +693,9 @@ func DeleteStopTask(ctx context.Context, taskID string) error {
 
 // DeleteActiveTasks returns a list of active deletion tasks started via DeleteRunTask().
 func DeleteActiveTasks(ctx context.Context) ([]*logstorage.DeleteTask, error) {
+	if externalStorage != nil {
+		return externalStorage.DeleteActiveTasks(ctx)
+	}
 	if localStorage != nil {
 		return localStorage.DeleteActiveTasks(ctx)
 	}
@@ -674,6 +704,9 @@ func DeleteActiveTasks(ctx context.Context) ([]*logstorage.DeleteTask, error) {
 
 // GetTenantIDs returns sorted tenantIDs on the [start..end] time range.
 func GetTenantIDs(ctx context.Context, start, end int64) ([]logstorage.TenantID, error) {
+	if externalStorage != nil {
+		return externalStorage.GetTenantIDs(ctx, start, end)
+	}
 	if localStorage != nil {
 		return localStorage.GetTenantIDs(ctx, start, end)
 	}
