diff --git a/go.mod b/go.mod
index ab90203..1530c91 100644
--- a/go.mod
+++ b/go.mod
@@ -2,6 +2,8 @@ module github.com/VictoriaMetrics/VictoriaTraces
 
 go 1.26.3
 
+replace github.com/VictoriaMetrics/VictoriaLogs => ../VictoriaLogs
+
 require (
 	github.com/VictoriaMetrics/VictoriaLogs v1.121.1-0.20260414190124-77df0c04d532 // v1.50.0
 	github.com/VictoriaMetrics/VictoriaMetrics v1.140.1-0.20260414051809-8a20ccf21db7
