diff --git a/go.mod b/go.mod
index 4b5b198..ee0ef75 100644
--- a/go.mod
+++ b/go.mod
@@ -4,6 +4,8 @@ go 1.26.2
 
 replace github.com/VictoriaMetrics/VictoriaMetrics => github.com/VictoriaMetrics/VictoriaMetrics v1.136.1-0.20260225205418-cd2026e4308c
 
+replace github.com/VictoriaMetrics/VictoriaLogs => ../VictoriaLogs
+
 require (
 	github.com/VictoriaMetrics/VictoriaLogs v1.47.1-0.20260225221819-a408207c2242
 	github.com/VictoriaMetrics/VictoriaMetrics v1.135.0
