diff --git a/app/vtstorage/main.go b/app/vtstorage/main.go
index cc07666..c670511 100644
--- a/app/vtstorage/main.go
+++ b/app/vtstorage/main.go
@@ -565,6 +565,10 @@ func isTimestampOutOfRetention(timestamp int64) bool {
 
 // RunQuery runs the given qctx and calls writeBlock for the returned data blocks
 func RunQuery(qctx *logstorage.QueryContext, writeBlock logstorage.WriteDataBlockFunc) error {
+	if externalStorage != nil {
+		return externalStorage.RunQuery(qctx, writeBlock)
+	}
+
 	qOpt, offset, limit := qctx.Query.GetLastNResultsQuery()
 	if qOpt != nil {
 		qctxOpt := qctx.WithQuery(qOpt)
@@ -585,6 +589,9 @@ func RunQuery(qctx *logstorage.QueryContext, writeBlock logstorage.WriteDataBloc
 //
 // If the filter isn't empty, then only the field names containing the filter substing are returned.
 func GetFieldNames(qctx *logstorage.QueryContext, filter string) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetFieldNames(qctx, filter)
+	}
 	if localStorage != nil {
 		return localStorage.GetFieldNames(qctx, filter)
 	}
@@ -597,6 +604,9 @@ func GetFieldNames(qctx *logstorage.QueryContext, filter string) ([]logstorage.V
 //
 // If limit > 0, then up to limit unique values are returned.
 func GetFieldValues(qctx *logstorage.QueryContext, fieldName, filter string, limit uint64) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetFieldValues(qctx, fieldName, filter, limit)
+	}
 	if localStorage != nil {
 		return localStorage.GetFieldValues(qctx, fieldName, filter, limit)
 	}
@@ -607,6 +617,9 @@ func GetFieldValues(qctx *logstorage.QueryContext, fieldName, filter string, lim
 //
 // If the filter is non-empty, then only the field names containing the filter substring are returned.
 func GetStreamFieldNames(qctx *logstorage.QueryContext, filter string) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetStreamFieldNames(qctx, filter)
+	}
 	if localStorage != nil {
 		return localStorage.GetStreamFieldNames(qctx, filter)
 	}
@@ -619,6 +632,9 @@ func GetStreamFieldNames(qctx *logstorage.QueryContext, filter string) ([]logsto
 //
 // If limit > 0, then up to limit unique stream field values are returned.
 func GetStreamFieldValues(qctx *logstorage.QueryContext, fieldName, filter string, limit uint64) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetStreamFieldValues(qctx, fieldName, filter, limit)
+	}
 	if localStorage != nil {
 		return localStorage.GetStreamFieldValues(qctx, fieldName, filter, limit)
 	}
@@ -629,6 +645,9 @@ func GetStreamFieldValues(qctx *logstorage.QueryContext, fieldName, filter strin
 //
 // If limit > 0, then up to limit unique streams are returned.
 func GetStreams(qctx *logstorage.QueryContext, limit uint64) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetStreams(qctx, limit)
+	}
 	if localStorage != nil {
 		return localStorage.GetStreams(qctx, limit)
 	}
@@ -639,6 +658,9 @@ func GetStreams(qctx *logstorage.QueryContext, limit uint64) ([]logstorage.Value
 //
 // If limit > 0, then up to limit unique streamIDs are returned.
 func GetStreamIDs(qctx *logstorage.QueryContext, limit uint64) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetStreamIDs(qctx, limit)
+	}
 	if localStorage != nil {
 		return localStorage.GetStreamIDs(qctx, limit)
 	}
@@ -683,6 +705,9 @@ func DeleteActiveTasks(ctx context.Context) ([]*logstorage.DeleteTask, error) {
 
 // GetTenantIDs returns tenantIDs from the storage by the given start and end.
 func GetTenantIDs(ctx context.Context, start, end int64) ([]logstorage.TenantID, error) {
+	if externalStorage != nil {
+		return externalStorage.GetTenantIDs(ctx, start, end)
+	}
 	if localStorage != nil {
 		return localStorage.GetTenantIDs(ctx, start, end)
 	}
