diff --git a/lib/logstorage/syslog_parser.go b/lib/logstorage/syslog_parser.go
index 990528903..ed6e9802d 100644
--- a/lib/logstorage/syslog_parser.go
+++ b/lib/logstorage/syslog_parser.go
@@ -404,6 +404,9 @@ func (p *SyslogParser) parseRFC5424SDLine(s string) (string, bool) {
 			return s, false
 		}
 		i += n + 1
+		if i == len(s) {
+			return s, false
+		}
 
 		// Parse value
 		if s[i] == '"' {
