diff --git a/app/vlstorage/main.go b/app/vlstorage/main.go
index 01cd23063..a9c64f6c1 100644
--- a/app/vlstorage/main.go
+++ b/app/vlstorage/main.go
@@ -552,6 +552,10 @@ func (*Storage) MustAddRows(lr *logstorage.LogRows) {
 
 // RunQuery runs the given qctx and calls writeBlock for the returned data blocks
 func RunQuery(qctx *logstorage.QueryContext, writeBlock logstorage.WriteDataBlockFunc) error {
+	if externalStorage != nil {
+		return externalStorage.RunQuery(qctx, writeBlock)
+	}
+
 	qOpt, offset, limit := qctx.Query.GetLastNResultsQuery()
 	if qOpt != nil {
 		qctxOpt := qctx.WithQuery(qOpt)
@@ -566,6 +570,9 @@ func RunQuery(qctx *logstorage.QueryContext, writeBlock logstorage.WriteDataBloc
 
 // GetFieldNames executes qctx and returns field names seen in results.
 func GetFieldNames(qctx *logstorage.QueryContext) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetFieldNames(qctx)
+	}
 	if localStorage != nil {
 		return localStorage.GetFieldNames(qctx)
 	}
@@ -576,6 +583,9 @@ func GetFieldNames(qctx *logstorage.QueryContext) ([]logstorage.ValueWithHits, e
 //
 // If limit > 0, then up to limit unique values are returned.
 func GetFieldValues(qctx *logstorage.QueryContext, fieldName string, limit uint64) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetFieldValues(qctx, fieldName, limit)
+	}
 	if localStorage != nil {
 		return localStorage.GetFieldValues(qctx, fieldName, limit)
 	}
@@ -584,6 +594,9 @@ func GetFieldValues(qctx *logstorage.QueryContext, fieldName string, limit uint6
 
 // GetStreamFieldNames executes the given qctx and returns stream field names seen in results.
 func GetStreamFieldNames(qctx *logstorage.QueryContext) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetStreamFieldNames(qctx)
+	}
 	if localStorage != nil {
 		return localStorage.GetStreamFieldNames(qctx)
 	}
@@ -594,6 +607,9 @@ func GetStreamFieldNames(qctx *logstorage.QueryContext) ([]logstorage.ValueWithH
 //
 // If limit > 0, then up to limit unique stream field values are returned.
 func GetStreamFieldValues(qctx *logstorage.QueryContext, fieldName string, limit uint64) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetStreamFieldValues(qctx, fieldName, limit)
+	}
 	if localStorage != nil {
 		return localStorage.GetStreamFieldValues(qctx, fieldName, limit)
 	}
@@ -604,6 +620,9 @@ func GetStreamFieldValues(qctx *logstorage.QueryContext, fieldName string, limit
 //
 // If limit > 0, then up to limit unique streams are returned.
 func GetStreams(qctx *logstorage.QueryContext, limit uint64) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetStreams(qctx, limit)
+	}
 	if localStorage != nil {
 		return localStorage.GetStreams(qctx, limit)
 	}
@@ -614,6 +633,9 @@ func GetStreams(qctx *logstorage.QueryContext, limit uint64) ([]logstorage.Value
 //
 // If limit > 0, then up to limit unique streamIDs are returned.
 func GetStreamIDs(qctx *logstorage.QueryContext, limit uint64) ([]logstorage.ValueWithHits, error) {
+	if externalStorage != nil {
+		return externalStorage.GetStreamIDs(qctx, limit)
+	}
 	if localStorage != nil {
 		return localStorage.GetStreamIDs(qctx, limit)
 	}
@@ -624,6 +646,10 @@ func GetStreamIDs(qctx *logstorage.QueryContext, limit uint64) ([]logstorage.Val
 //
 // The taskID and timestamp are tracked in the list of tasks returned by DeleteActiveTasks().
 func DeleteRunTask(ctx context.Context, taskID string, timestamp int64, tenantIDs []logstorage.TenantID, f *logstorage.Filter) error {
+	if externalStorage != nil {
+		return externalStorage.DeleteRunTask(ctx, taskID, timestamp, tenantIDs, f)
+	}
+
 	logger.Infof("starting deleting logs for task_id=%q, filter=%q, tenantIDs=%s", taskID, f, tenantIDs)
 
 	if localStorage != nil {
@@ -634,6 +660,10 @@ func DeleteRunTask(ctx context.Context, taskID string, timestamp int64, tenantID
 
 // DeleteStopTask stops delete task with the given taskID.
 func DeleteStopTask(ctx context.Context, taskID string) error {
+	if externalStorage != nil {
+		return externalStorage.DeleteStopTask(ctx, taskID)
+	}
+
 	logger.Infof("stopping delete task with task_id=%q", taskID)
 
 	var err error
@@ -650,6 +680,9 @@ func DeleteStopTask(ctx context.Context, taskID string) error {
 
 // DeleteActiveTasks returns a list of active deletion tasks started via DeleteRunTask().
 func DeleteActiveTasks(ctx context.Context) ([]*logstorage.DeleteTask, error) {
+	if externalStorage != nil {
+		return externalStorage.DeleteActiveTasks(ctx)
+	}
 	if localStorage != nil {
 		return localStorage.DeleteActiveTasks(ctx)
 	}
@@ -658,6 +691,9 @@ func DeleteActiveTasks(ctx context.Context) ([]*logstorage.DeleteTask, error) {
 
 // GetTenantIDs returns tenantIDs from the storage by the given start and end.
 func GetTenantIDs(ctx context.Context, start, end int64) ([]logstorage.TenantID, error) {
+	if externalStorage != nil {
+		return externalStorage.GetTenantIDs(ctx, start, end)
+	}
 	if localStorage != nil {
 		return localStorage.GetTenantIDs(ctx, start, end)
 	}
